### PR TITLE
adding runai cli command annotation

### PR DESCRIPTION
**Pull request by moshe-kabala**

### CODE REVIEW

File: charts/mpijob/templates/mpi-job.yaml [hunk 1]
```diff
@@ -10,6 +10,7 @@ metadata:
     image: "{{ .Values.image }}"
     totalGPUs: "{{ .Values.totalGpus }}"
     totalGPUsMemory: "{{ .Values.totalGpusMemory }}"
+    runai-cli-command: {{ .Values.runaiCliCommand | quote }}
```
Comment (yodarshafrir1) [inline charts/mpijob/templates/mpi-job.yaml:13]: why did you add it to both mpijob annotation and pod annotation?

File: charts/runai/templates/runai-job.yaml [hunk 2]
```diff
@@ -35,10 +35,11 @@ metadata:
     {{ $key }}: {{ $val | quote }}
     {{- end }}
   {{- end}}
-  {{- if .Values.user }}
   annotations:
+    runai-cli-command: {{ .Values.runaiCliCommand | quote }}
```
Comment (yodarshafrir1) [inline charts/runai/templates/runai-job.yaml:39]: why did you add it to both runaijob annotation and pod annotation?

File: cmd/job/describe_job.go [hunk 3]
```diff
@@ -344,3 +346,10 @@ func getPriorityClass(job trainer.TrainingJob) string {
 
 	return pc
 }
+
+func getCommandLine(job trainer.TrainingJob) ( commandLine string ) {
```
Comment (yodarshafrir1) [inline cmd/job/describe_job.go:350]: if no command line write `N/A` or something similiar

File: cmd/job/submit/submit.go [hunk 4]
```diff
@@ -80,6 +83,7 @@ type submitArgs struct {
 	Memory              string   `yaml:"memory,omitempty"`
 	MemoryLimit         string   `yaml:"memoryLimit,omitempty"`
 	EnvironmentVariable []string `yaml:"environment,omitempty"`
+	RunaiCliCommand		string   `yaml:"runaiCliCommand,omitempty"`
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit.go:86]: `cliCommand`

File: cmd/job/submit/submit_mpi.go [hunk 5]
```diff
@@ -102,6 +102,7 @@ func NewRunaiSubmitMPIJobCommand() *cobra.Command {
 
 	fbg := flags.NewFlagsByGroups(command)
 	submitArgs.addCommonFlags(fbg)
+	submitArgs.addRunaiCliCommand()
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_mpi.go:105]: you are calling this method from all the job kinds.
better use a common method like: `addCommonSubmit` and from this method call `addCommonFlags` and `addRunaiCliCommand`

File: cmd/job/submit/submit.go [hunk 6]
```diff
@@ -123,6 +127,10 @@ func (s submitArgs) check() error {
 	return nil
 }
 
+func(submitArgs *submitArgs) addRunaiCliCommand() {
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit.go:130]: why do we need the runai in the name?

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]

File: charts/mpijob/templates/mpi-job.yaml (hunk 1 above)
Comment (yodarshafrir1) [reply charts/mpijob/templates/mpi-job.yaml:13]: better stick with job's annotation

Comment (moshe-kabala) [reply charts/mpijob/templates/mpi-job.yaml:13]: In the scheduler, I extract from the pod but in the researcher UI from the job.

File: charts/runai/templates/runai-job.yaml (hunk 2 above)
Comment (moshe-kabala) [reply charts/runai/templates/runai-job.yaml:39]: In the scheduler, I extract from the pod but in the researcher UI from the job.

File: cmd/job/submit/submit.go (hunk 6 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:130]: fixed

File: cmd/job/submit/submit_mpi.go (hunk 5 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit_mpi.go:105]: fixed

File: cmd/job/submit/submit.go (hunk 4 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:86]: fixed

File: cmd/job/describe_job.go (hunk 3 above)
Comment (moshe-kabala) [reply cmd/job/describe_job.go:350]: fixed

File: charts/mpijob/templates/mpi-job.yaml (hunk 1 above)
Comment (yodarshafrir1) [reply charts/mpijob/templates/mpi-job.yaml]: why not do the same in both?

**Comment by yodarshafrir1** [comment]
Instead of putting the same label on both annotations of job and pods, can we stick to one of them? (the job is better)

File: charts/mpijob/templates/mpi-job.yaml (hunk 1 above)
Comment (moshe-kabala) [reply charts/mpijob/templates/mpi-job.yaml]: As agreed I will remain with the 'job' annotation

File: charts/runai/templates/runai-job.yaml (hunk 2 above)
Comment (moshe-kabala) [reply charts/runai/templates/runai-job.yaml]: As agreed I will remain with the 'job' annotation

File: cmd/job/describe_job.go [hunk 7]
```diff
@@ -344,3 +346,15 @@ func getPriorityClass(job trainer.TrainingJob) string {
 
 	return pc
 }
+
+func getCliCommand(job trainer.TrainingJob) string {
+	return stringWithDefault(job.CliCommand(), "N/A")
+}
+
+func stringWithDefault(str string, def string) string {
```
Comment (omer-dayan) [inline cmd/job/describe_job.go:354]: Its better call "getStringOrDefault", because "with" is not really what we do

Comment (omer-dayan) [reply cmd/job/describe_job.go:354]: Do we really want seperate function for this?

File: cmd/job/submit/submit.go [hunk 8]
```diff
@@ -80,6 +83,7 @@ type submitArgs struct {
 	Memory              string   `yaml:"memory,omitempty"`
 	MemoryLimit         string   `yaml:"memoryLimit,omitempty"`
 	EnvironmentVariable []string `yaml:"environment,omitempty"`
+	CliCommand		     string  `yaml:"cliCommand,omitempty"`
```
Comment (omer-dayan) [inline cmd/job/submit/submit.go:86]: Why is it public if its only used within this package?

File: cmd/job/submit/submit.go [hunk 9]
```diff
@@ -123,6 +127,15 @@ func (s submitArgs) check() error {
 	return nil
 }
 
+func(submitArgs *submitArgs) addCommonSubmit(fbg flags.FlagsByGroups) {
+	submitArgs.addCommonFlags(fbg)
+	submitArgs.addCliCommand()
+}
+
+func(submitArgs *submitArgs) addCliCommand() {
+	submitArgs.CliCommand = strings.Join( os.Args , " ")
```
Comment (omer-dayan) [inline cmd/job/submit/submit.go:136]: NO NO its bad to save it that way, attacker can interfere our YAML file

It should be saved as BASE64

File: cmd/util/gpu.go [hunk 10]
```diff
@@ -32,6 +32,7 @@ const (
 	DeprecatedNVIDIAGPUResourceName    = "alpha.kubernetes.io/nvidia-gpu"
 	PodGroupRequestedGPUs              = "runai-podgroup-requested-gpus"
 	PodGroupRequestedGPUsMemory        = "runai-podgroup-requested-gpus-memory"
+	CliCommand        				   = "runai-cli-command"
```
Comment (omer-dayan) [inline cmd/util/gpu.go:35]: Its not related to GPU, move it to consts.go

**Review by omer-dayan — COMMENTED** [review]
Liked it,
Please notice its super important to save the CLI Command as BASE64 and not raw text

Also: What about the default template? The CLI command may be mis-leading by its own

File: cmd/job/describe_job.go (hunk 7 above)
Comment (moshe-kabala) [reply cmd/job/describe_job.go:354]: deleted

File: cmd/job/submit/submit.go (hunk 8 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:86]: It private now

File: cmd/util/gpu.go (hunk 10 above)
Comment (moshe-kabala) [reply cmd/util/gpu.go:35]: moved to the const file

File: cmd/job/submit/submit.go (hunk 9 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:136]: It is OK we are using  quote (in the template)

Comment (omer-dayan) [reply cmd/job/submit/submit.go:136]: Yes its doing escaping but we may lose data when escaping, For example: 

The command : "runai submit -i ubuntu -g 0 --dry-run -- echo \n\t hello"
Becomes : "runai submit -i ubuntu -g 0 --dry-run -- echo nt hello"

**Review by omer-dayan — APPROVED** [review]

**Pull request merged**